src/extensions/extension-loader/extension-loader.ts

```diff
@@ -312,35 +317,58 @@ export class ExtensionLoader {
 
             if (!LensExtensionClass) {
               this.nonInstancesByName.add(extension.manifest.name);
-              continue;
+
+              return null;
             }
 
             const instance = this.dependencies.createExtensionInstance(
               LensExtensionClass,
               extension,
             );
 
-            const loaded = instance.enable(register).catch((err) => {
-              logger.error(`${logModule}: failed to enable`, { ext: extension, err });
-            });
-
-            loadingExtensions.push({
+            return {
+              extId,
+              instance,
               isBundled: extension.isBundled,
-              loaded,
-            });
-            this.instances.set(extId, instance);
+              activated: instance.activate(),
+            };
           } catch (err) {
             logger.error(`${logModule}: activation extension error`, { ext: extension, err });
           }
         } else if (!extension.isEnabled && alreadyInit) {
           this.removeInstance(extId);
         }
-      }
-    }, {
-      fireImmediately: true,
+
+        return null;
+      })
+      // Remove null values
+      .filter(extension => Boolean(extension));
+
+    // We first need to wait until each extension's `onActivate` is resolved,
+    // as this might register new catalog categories. Afterwards we can safely .enable the extension.
+    await Promise.all(extensions.map(extension => extension.activated));
+
+    // Return ExtensionLoading[]
+    return extensions.map(extension => {
+      const loaded = extension.instance.enable(register).catch((err) => {
+        logger.error(`${logModule}: failed to enable`, { ext: extension, err });
+      });
+
+      this.instances.set(extension.extId, extension.instance);
+
+      return {
+        isBundled: extension.isBundled,
+        loaded,
+      };
     });
+  }
+
+  protected autoInitExtensions(register: (ext: LensExtension) => Promise<Disposer[]>) {
+    // Setup reaction to load extensions on JSON changes
+    reaction(() => this.toJSON(), installedExtensions => this.loadExtensions(installedExtensions, register));
```

Comment:
`this.toJSON()` seems like always return new object/value, should we limit amount of reactions maybe with 
`{fireImmediately: true, equals: comparer.structural}`?

Reply:
I removed `fireImmediately: true`, since I want to load extensions once when this is called, but return a promise, which is resolved after extensions have been loaded.

The reaction tracks one observable:  `this.extensions` of `ExtensionLoader`. It seems that `this.extensions` is only updated if there is some change, so this should work correctly even without `comparer.structural`.